iptables/table.go

```diff
@@ -795,17 +789,25 @@ func (t *Table) readHashesAndRulesFrom(r io.ReadCloser) (hashes map[string][]str
 		}
 		hashes[chainName] = append(hashes[chainName], hash)
 
-		// Either we've explicitly inserted a rule into this chain or it's one of the top-level chains that we may have hooked.
-		// Store off the complete contents of the rules so that we can generate precise deletes later if we need to.
-		if _, ok := t.chainToInsertedRules[chainName]; t.topLevelChains.Contains(chainName) || ok {
+		// Not our chain so cache the full rule in case we need to generate deletes later on.
+		// After scanning the input, we prune any chains of full rules that do not contain inserts.
+		if !t.ourChainsRegexp.MatchString(chainName) {
 			fullRule := string(line)
 			rules[chainName] = append(rules[chainName], fullRule)
+			nonCalicoChainHasInserts[chainName] = true
```

Comment:
I think you need to check whether this particular rule is actually an insert or not before setting this flag.  Right now, IIUC, you'll set it for every non-Calico chain.

Reply:
I think you can piggy-back on the logic just above this, Just add `nonCalicoChainHasInserts[chainName] = true` to both arms of the conditional above (where we get the hash or set hash to "OLD INSERT RULE").  Since the above conditional fires for all chains, I'd rename `nonCalicoChainHasInserts` to `chainHasCalicoRule`, but the logic below will carry on working with that change.